README

    ╔╗ ┬┬  ┬  ┌┐ ┌─┐┌─┐┬─┐┌┬┐
    ╠╩╗││  │  ├┴┐│ │├─┤├┬┘ ││
    ╚═╝┴┴─┘┴─┘└─┘└─┘┴ ┴┴└──┴┘
             Standby Screen


   "One of the great things about the Nokia N9 is the AMOLED screen with
    its low power mode. Unfortunately, it only shows the current time,
    and not the current date. Billboard is here to fix this - and add
    some more goodies along the way!"       -- me, circa 2012


This is the open source release of my N9 application, Billboard.
A short overview of the structure of what's inside:

 - billboardd/
   The daemon source code and all the core code (including formatting,
   rendering and contextkit/gconf subscription).

 - billboard-ui/
   The QML UI for the "app", to configure the daemon. Communicates with
   the daemon via shared GConf keys.

 - css/
   Stylesheets for making the operator logo not be cropped.

 - debian/
   Packaging files and postinst/postrm files for patching the stylesheet
   and installing stuff, as well as aegis files (for organizer/calendar).

 - include/
   Common configuration key defaults used by both the UI and daemon.

 - rendertest/
   Desktop GUI utility for testing text formatting and rendering.


If you didn't use "git clone --recursive" when closing this Git repo,
you have to initialize the submodule (to get the translations; this
does not apply to the source tarball, as it already includes i18n):

    git submodule init
    git submodule update


To build for the N9 in Scratchbox use:

    make -C billboard-ui/i18n release
    dpkg-buildpackage -rfakeroot -b


Useful scripts to be used together with Billboard may be found here:

    https://github.com/harmattan/billboard-scripts


License: GNU General Public License, version 3 or later
Support thread: http://talk.maemo.org/showthread.php?t=84507
Website: http://thp.io/2012/billboard/


debian/changelog
billboard (2.2.0) harmattan; urgency=low

  * [Elliot Wolk] aegis: Add D-Bus credentials (billboardd, billboard-cli)
  * [Elliot Wolk] dbus: Print message when D-Bus service registration fails
  * [Thomas Perl] Releasing: Read all version numbers from debian/changelog

 -- Thomas Perl <m@thp.io>  Wed, 08 Jul 2015 10:04:00 +0200

billboard (2.0.1) harmattan; urgency=low

  * i18n: Updated translations from billboard-i18n
  * tklock: call contextChanged() on tklock dbus event
  * dbus: interface to implement 'qdbus io.thp.billboard / render'
  * billboard-cli: install a wrapper around dbus-send

 -- Thomas Perl <m@thp.io>  Wed, 06 May 2015 15:33:10 +0200

billboard (2.0.0) harmattan; urgency=low

  * Open Source release under the GPLv3

 -- Thomas Perl <m@thp.io>  Thu, 04 Sep 2014 15:06:00 +0100

billboard (1.1.2) harmattan; urgency=low

  * New translations: ru
  * Updated translations: it

 -- Thomas Perl <m@thp.io>  Sun, 23 Mar 2014 15:04:24 +0100

billboard (1.1.1) harmattan; urgency=low

  * Refreshed icon for the configuration utility
  * Add UI elements for cell signal strength and bars
  * Add {short-month} element for Jan, Feb, ...
  * New translations: el, it, sk
  * Updated translations: zh

 -- Thomas Perl <m@thp.io>  Thu, 03 Oct 2013 15:35:01 +0200

billboard (1.1.0) harmattan; urgency=low

  * Configuration utility can now insert <<>>, {} and {{}}
  * Only limit embedded image size when dithering is enabled
  * Other logos and dithered images can now be up to 200x120 in size
  * i18n Support in QML configuration utility
  * New translations: bg, da, de, en, hi, nl, pt, zh

 -- Thomas Perl <m@thp.io>  Mon, 06 May 2013 17:36:46 +0200

billboard (1.0.9) harmattan; urgency=low

  * Fixed MeeCast icon (use <<{meecast-icon-src}>>)
  * New formatter that supports nested {} expressions
  * LPM-optimized image dithering for better display quality
  * Limit size of embedded images to avoid dithering lock-up
  * Image dithering using <</path/to/image.png#>> (# at end of path)

 -- Thomas Perl <m@thp.io>  Thu, 04 Apr 2013 17:09:04 +0200

billboard (1.0.8) harmattan; urgency=low

  * MeeCast icon display support ({meecast-icon})
  * Wazapp online status now accessible from GUI
  * USSD Balance now accessible from GUI
  * Signal info ({cell-signal-bars} and {cell-signal-strength})
  * Alarm present info ({alarm-present})
  * Support for mixing images and text in the same line
  * Center embedded images vertically (and don't scale them)
  * Conflits: and Replaces: billboardlite (Demo Version)
  * Update version information in the QML UI

 -- Thomas Perl <m@thp.io>  Thu, 13 Dec 2012 12:32:03 +0100

billboard (1.0.7) harmattan; urgency=low

  * New rendering engine
  * Inline color changing ({{red}}, etc...)
  * Added week number field ({week})
  * Battery charge support (as graphical bar)
  * Colored battery bar ({color-battery-bar}, colored when low)
  * Add support for wazapp online status (Wazapp.Online)
  * Add support for GConf Key Monitoring and USSD Balance
  * Add support for showing a battery icon ({battery-icon})
  * Script output is now interpreted as UTF-8 encoded string
  * Allow up to 20 lines of text (up from 7 previously)

 -- Thomas Perl <m@thp.io>  Mon, 01 Oct 2012 23:27:09 +0200

billboard (1.0.6) harmattan; urgency=low

  * ContextKit integration with MeeCast
  * Support for displaying only the month number

 -- Thomas Perl <m@thp.io>  Sat, 01 Sep 2012 10:55:41 +0200

billboard (1.0.5) harmattan; urgency=low

  * Display IP address(es)

 -- Thomas Perl <m@thp.io>  Wed, 01 Aug 2012 03:01:07 +0200

billboard (1.0.4) harmattan; urgency=low

  * Customizable text size
  * Customizable font family

 -- Thomas Perl <m@thp.io>  Mon, 23 Jul 2012 19:52:09 +0200

billboard (1.0.3) harmattan; urgency=low

  * Fix problems for users who have already upgraded to PR1.3

 -- Thomas Perl <m@thp.io>  Mon, 09 Jul 2012 19:40:18 +0200

billboard (1.0.2) harmattan; urgency=low

  * Use system-aligned timer and update every 30-60 minutes
    (if no other events take place, this fixes date updates)
  * Update postinst/prerm and restart sysuid directly
  * Add support for showing calendar events (today/upcoming)
  * Add support for more date formatting options (DoM, weekday, ..)
  * Fixed a problem with center-aligned text and other logos

 -- Thomas Perl <m@thp.io>  Thu, 07 Jun 2012 20:54:23 +0200

billboard (1.0.1) harmattan; urgency=low

  * Override other operator logos (and restore on quit).
  * Draw other operator logo onto the billboard, too.
  * Better layout (minimum required size), right-aligned.
  * Elide text that is too wide for the screen.
  * Limit other operator logo size to 120x120px.
  * Option to disable billboard through the UI.
  * Option to show/hide the other operator logo.
  * New settings UI.
  * New text alignment option (left, right center).
  * More robust replacement of sysuid's .css file.

 -- Thomas Perl <m@thp.io>  Sat, 26 May 2012 15:12:10 +0200

billboard (1.0.0) harmattan; urgency=low

  * Initial Release.

 -- Thomas Perl <m@thp.io>  Mon, 14 May 2012 21:03:31 +0200


billboardd/daemon.h

/**
 * Billboard - Low Power Mode Standby Screen for the N9
 * Webpage: http://thp.io/2012/billboard/
 * Copyright (C) 2012, 2013, 2014 Thomas Perl <thp.io/about>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 **/

#ifndef BILLBOARDD_DAEMON_H
#define BILLBOARDD_DAEMON_H

#include <QtCore>
#include <QtGui>
#include <GConfItem>
#include <QSystemAlignedTimer>

QTM_USE_NAMESPACE

#include "../include/defaults.h"
#include "context.h"
#include "gconfmon.h"
#include "organizer.h"
#include "network.h"
#include "tklockinterface.h"
#include "volumeinterface.h"
#include "batteryicon.h"

#include "renderer.h"
#include "formatter.h"

#define LOGO_KEY "/desktop/meego/screen_lock/low_power_mode/operator_logo"
#define CONFIG_KEY "/apps/billboard"

#define EVERY_HALF_HOUR (60*30)
#define EVERY_HOUR (60*60)

#define BATTERY_WARNING_FRAC (.3)
#define BATTERY_CRITICAL_FRAC (.15)

class Daemon : public QObject {
    Q_OBJECT

    public:
        Daemon(QString path, QObject *parent=NULL)
            : QObject(parent),
              file_path(path),
              filename_a(file_path.filePath("frame_a.png")),
              filename_b(file_path.filePath("frame_b.png")),
              quitting(false),
              operator_logo(NULL),
              config_other_logo(NULL),
              config_other_enabled(NULL),
              config_text(NULL),
              config_color(NULL),
              config_alignment(NULL),
              config_enabled(NULL),
              config_size(NULL),
              config_font(NULL),
              enabled(false),
              context(NULL),
              gconfmon(NULL),
              organizer(),
              tklock(),
              volume(),
              updateTimer(),
              sysAlignedTimer(),
              lastRender(""),
              templateText("")
        {
            operator_logo = new GConfItem(LOGO_KEY, this);
            config_other_logo = new GConfItem(CONFIG_KEY "/other_logo", this);
            config_other_enabled = new GConfItem(CONFIG_KEY "/other_enabled", this);
            config_text = new GConfItem(CONFIG_KEY "/text", this);
            config_color = new GConfItem(CONFIG_KEY "/color", this);
            config_alignment = new GConfItem(CONFIG_KEY "/alignment", this);
            config_enabled = new GConfItem(CONFIG_KEY "/enabled", this);
            config_size = new GConfItem(CONFIG_KEY "/size", this);
            config_font = new GConfItem(CONFIG_KEY "/font", this);

            QObject::connect(operator_logo, SIGNAL(valueChanged()),
                    this, SLOT(operatorLogoChanged()));
            QObject::connect(config_other_logo, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));
            QObject::connect(config_other_enabled, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));
            QObject::connect(config_text, SIGNAL(valueChanged()),
                    this, SLOT(textChanged()));
            QObject::connect(config_color, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));
            QObject::connect(config_alignment, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));
            QObject::connect(config_enabled, SIGNAL(valueChanged()),
                    this, SLOT(enabledChanged()));
            QObject::connect(config_size, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));
            QObject::connect(config_font, SIGNAL(valueChanged()),
                    this, SLOT(contextChanged()));

            context = new Context(this);
            QObject::connect(context, SIGNAL(changed()),
                    this, SLOT(contextChanged()));

            gconfmon = new GConfMonitor(this);
            QObject::connect(gconfmon, SIGNAL(changed()),
                    this, SLOT(contextChanged()));

            updateTimer.setInterval(500);
            updateTimer.setSingleShot(true);

            QObject::connect(&updateTimer, SIGNAL(timeout()),
                    this, SLOT(render()));

            sysAlignedTimer.setMinimumInterval(EVERY_HALF_HOUR);
            sysAlignedTimer.setMaximumInterval(EVERY_HOUR);
            sysAlignedTimer.setSingleShot(false);
            sysAlignedTimer.start();

            QObject::connect(&sysAlignedTimer, SIGNAL(timeout()),
                    this, SLOT(sysAlignedUpdate()));

            QObject::connect(&tklock, SIGNAL(locked()),
                    this, SLOT(contextChanged()));

            enabledChanged();
            textChanged();
            render(true);
        }

        ~Daemon() {
            quitting = true;
            resetOperatorLogo();
        }

    private:
        QDir file_path;
        QString filename_a;
        QString filename_b;
        bool quitting;
        GConfItem *operator_logo;
        GConfItem *config_other_logo;
        GConfItem *config_other_enabled;
        GConfItem *config_text;
        GConfItem *config_color;
        GConfItem *config_alignment;
        GConfItem *config_enabled;
        GConfItem *config_size;
        GConfItem *config_font;
        bool enabled;

        Context *context;
        GConfMonitor *gconfmon;
        Organizer organizer;
        TklockInterface tklock;
        VolumeInterface volume;

        QTimer updateTimer;
        QSystemAlignedTimer sysAlignedTimer;
        QString lastRender;
        QString templateText;

        QList<QString> getInterestingProperties() {
            QSet<QString> props;

            Formatter::format(templateText, templatemap(), &props);

            if (props.contains("ip")) {
                props << "network-name";
            }

            if (props.contains("song-volume-bar")) {
                props << "volume";
                props << "song";
            }
            if (props.contains("volume-bar")) {
                props << "volume";
            }

            if (props.contains("battery-bar")) {
                props << "battery";
            }
            if (props.contains("color-battery-bar")) {
                props << "battery";
            }
            if (props.contains("battery-icon")) {
                props << "battery";
            }

            return props.toList();
        }

        QMap<QString,QString> templatemap() {
            QDate now = QDate::currentDate();

            QString shortDate = now.toString(Qt::DefaultLocaleShortDate);
            QString longDate = now.toString(Qt::DefaultLocaleLongDate);

            QMap<QString,QString> tmpl;
            tmpl["date"] = shortDate;
            tmpl["full-date"] = longDate;

            tmpl["week"] = QString::number(now.weekNumber());
            tmpl["weekday"] = now.toString("dddd");
            tmpl["day"] = now.toString("d");
            tmpl["month"] = now.toString("MMMM");
            tmpl["short-month"] = now.toString("MMM");
            tmpl["month-number"] = now.toString("M");
            tmpl["year"] = now.toString("yyyy");
            tmpl["ip"] = get_ip_addresses();

            tmpl["events"] = organizer.getUpcomingEvents();
            tmpl["events-today"] = organizer.getUpcomingEvents(0);

            tmpl.unite(context->properties());
            tmpl.unite(gconfmon->properties());

            double battery = 0.01 * tmpl["battery"].toDouble();
            tmpl["battery-bar"] = "{{=" + QString::number(battery) + "}}";
            tmpl["battery-icon"] = "<<" + BatteryIcon::fromPercentage(battery) + ">>";

            // Colored battery bar - yellow and red
            if (battery < BATTERY_CRITICAL_FRAC) {
                tmpl["color-battery-bar"] = "{{red}}" + tmpl["battery-bar"] + "{{}}";
            } else if (battery < BATTERY_WARNING_FRAC) {
                tmpl["color-battery-bar"] = "{{yellow}}" + tmpl["battery-bar"] + "{{}}";
            } else {
                tmpl["color-battery-bar"] = tmpl["battery-bar"];
            }

            tmpl["battery-bar"] = "{{=" + QString::number((double)(tmpl["battery"].toDouble()*.01)) + "}}";

            tmpl["volume"] = QString::number((int)(volume.volume()*100.));
            tmpl["volume-bar"] = "{{=" + QString::number(volume.volume()) + "}}";

            // Only show volume bar when something is being played back
            if (tmpl["song"] != "") {
                tmpl["song-volume-bar"] = tmpl["volume-bar"];
            } else {
                tmpl["song-volume-bar"] = "";
            }

            return tmpl;
        }

        void resetOperatorLogo() {
            operator_logo->set(config_other_logo->value(""));
        }

    public slots:
        void enabledChanged() {
            bool newEnabled = config_enabled->value(BILLBOARD_CONFIG_ENABLED).toBool();

            if (enabled != newEnabled) {
                enabled = newEnabled;

                if (enabled) {
                    /* Re-subscribe to contextkit subscriptions */
                    textChanged();

                    /* Render twice, just to be sure it picks the right file? */
                    render(true);
                    render(true);
                } else {
                    /* Unsubscribe from all contextkit subscriptions */
                    context->subscribeTo(QList<QString>());

                    /* Unsubscribe from all gconf monitor subscriptions */
                    gconfmon->subscribeTo(QList<QString>());

                    /* Unsubscribe from volume notifications */
                    volume.disconnect();

                    /* Reset operator logo to "other" logo (if any) */
                    resetOperatorLogo();
                }
            }
        }

        void contextChanged() {
            updateTimer.start();
        }

        void textChanged() {
            if (!enabled) {
                return;
            }

            templateText = config_text->value(BILLBOARD_CONFIG_TEXT).toString();
            QList<QString> interestingProperties = getInterestingProperties();
            context->subscribeTo(interestingProperties);
            gconfmon->subscribeTo(interestingProperties);

            if (interestingProperties.contains("volume")) {
                QObject::connect(&volume, SIGNAL(changed()),
                        this, SLOT(contextChanged()));
            } else {
                volume.disconnect();
            }

            updateTimer.start();
        }

        void operatorLogoChanged() {
            if (quitting) {
                /* Ignore operator logo changes if we are quitting */
                return;
            }

            QString filename = operator_logo->value("").toString();
            if (filename != filename_a && filename != filename_b) {
                config_other_logo->set(filename);

                /* Render twice to make sure the logo is updated and seen */
                render(true);
                render(true);
            }
        }

        void sysAlignedUpdate()
        {
            render();
        }

        void render(bool force=false)
        {
            sysAlignedTimer.wokeUp();

            if (!enabled) {
                /* Do not render anything, even when forced */
                return;
            }

            updateTimer.stop();

            QString text = Formatter::format(templateText, templatemap(), NULL);

            if (!force && lastRender == text) {
                /* Don't update when not necessary */
                return;
            }

            int flags = 0;

            QString alignment = config_alignment->value(BILLBOARD_CONFIG_ALIGNMENT).toString();
            if (alignment == "left") {
                flags |= Qt::AlignLeft;
            } else if (alignment == "center") {
                flags |= Qt::AlignHCenter;
            } else {
                flags |= Qt::AlignRight;
            }

            QImage other;
            if (config_other_enabled->value(BILLBOARD_CONFIG_OTHER_ENABLED).toBool()) {
                other = QImage(config_other_logo->value("").toString());
            }

            QColor color = QColor(config_color->value(BILLBOARD_CONFIG_COLOR).toString());
            QFont font = QFont(config_font->value(BILLBOARD_CONFIG_FONT).toString(),
                    config_size->value(BILLBOARD_CONFIG_SIZE).toInt());

            QImage rendered = Renderer::render(text, flags, other, color, font);

            QString filename = ((operator_logo->value("") == filename_a)?
                    filename_b:filename_a);

            rendered.save(filename);

            operator_logo->set(filename);
        }
};

#endif
